internal/engine/compiler/engine.go

```diff
@@ -702,14 +702,23 @@ func (ce *callEngine) Definition() api.FunctionDefinition {
 
 // Call implements the same method as documented on wasm.ModuleEngine.
 func (ce *callEngine) Call(ctx context.Context, params ...uint64) (results []uint64, err error) {
+	return ce.call(ctx, params, nil)
+}
+
+// CallWithStack implements the same method as documented on wasm.ModuleEngine.
+func (ce *callEngine) CallWithStack(ctx context.Context, stack []uint64) (res []uint64, err error) {
```

Comment:
Supposedly the caller needs to know how many return values there are, would it make sense to accept the params and results as inputs? Changing the signature to:
```go
CallWithStack(ctx context.Context, params, results []uint64) error { ... }
```

Reply:
The caller doesn't _really_ need to know how many return values there are.
We could demand that they do.

This how I'm planing to use `CallWithStack`, vs. what I [currently do](https://github.com/ncruces/go-sqlite3/blob/a6226c3b31fe51636a28a7b188df95be76c68914/module.go#L207-L216):
```go
func (m *module) call(fn api.Function, params ...uint64) []uint64 {
	m.arg = append(m.arg[:0], params...)
	r, err := fn.CallWithStack(m.ctx, m.arg)
	if err != nil {
		// The module closed or panicked; release resources.
		m.vfs.Close()
		panic(err)
	}
	return r
}
```

This `call` method wraps all my `api.Function` uses; so `m.arg` is a `[]uint64` reused across all calls into the same `api.Module`. It eventually grows to hold enough space for the maximum number of args/results, and no more temporary allocations happen.

With your API, I'd have to do:
```go
func (m *module) call(fn api.Function, params ...uint64) []uint64 {
	resCount := len(fn.Definition().ResultTypes())
	if resCount > cap(m.arg) {
		m.arg = make([]uint64, resCount)
	}
	m.arg = append(m.arg[:0], params...)
	err := fn.CallWithStack(m.ctx, m.arg, m.arg[:resCount])
	if err != nil {
		// The module closed or panicked; release resources.
		m.vfs.Close()
		panic(err)
	}
	return m.arg[:resCount]
}
```

Slightly more convoluted, but maybe I pigeonholed into my own API.

Reply:
I agree the caller doesn't need to know, but if it doesn't care it can pass a buffer "large enough" (in the common case there will rarely be more than a few results, usually zero or one). If it needs a larger buffer it will be visible on a memory profile.

If we want to keep the returned `[]uint64` I think that's fine, but not mixing the input and output in a single parameter seems like a safer API. If the program really wants to use a single buffer we still can support it (it's the current implementation), but not forcing it would likely be less error prone and cover more use cases.

Reply:
I went with @codefromthecrypt's suggested API. TBH I don't have a strong preference.

The advantage of returning the result is that the `len` of the returned slice is the number of returned values and a middleman doesn't need to know.

How about we expose the private `call` method from this PR? Would that be a decent API? You can pass independent input/output slices, or the same slice (I'm doing both here). If the output slice does not have enough capacity for the result, a new one is returned. A little like `append` and `Append` methods of the standard lib.

I'd honestly like to see more people trying this in other codebases to be sure.

Reply:
I really don't like being different from the other API especially as this is an advanced one anyway. It seems we have more opportunities to optimize internally as well, when the shape matches the host side. Having something both unlike api.Function an also unlike api.GoXFunction seems far from ideal iotw.